packages/ngx-flicking/projects/ngx-flicking/src/lib/ngx-flicking.component.ts

```diff
@@ -2,286 +2,233 @@
 
-        if (emitter) {
-          emitter.emit(e);
-
-          if (eventName === 'visibleChange') {
-            const list = this.counter(this.panels.length * (this.flicking.getCloneCount() + 1));
-            const min = e.range.min;
-            const max = e.range.max;
-
-            const visibles = min >= 0
-              ? list.slice(min, max + 1)
-              : list.slice(0, max + 1).concat(list.slice(min));
-
-            this.triggerRenderChange(visibles);
-          }
-        }
+        if (emitter)
+        emitter.emit(e);
       });
     });
-  }
 
-  private triggerRenderChange(visibles: number[]) {
-    // visible panel is not changed
-    if (visibles.length === this.prevVisibles.length &&
-      visibles.every((v, i) => this.prevVisibles[i] === v)) {
-      return;
+    if (this.options.renderOnlyVisible) {
+      flicking.on(EVENTS.VISIBLE_CHANGE, this._reRender);
     }
+  }
 
-    this.prevVisibles = visibles;
+  private _checkPlugins() {
+    const flicking = this._nativeFlicking;
+    if (!flicking) return;
 
-    const l = this.panels.length;
-    const renderChangeEvent: RenderPanelChangeEvent = {
-      visibles: visibles.map(i => ({
-        key: i,
-        index: i % l,
-        data: this.panels[i % l]
-      }))
-    };
+    const { list, added, removed, prevList } = this._pluginsDiffer.update(this.plugins);
 
-    if (Promise) {
-      Promise.resolve()
-        .then(() => this.renderPanelChange.emit(renderChangeEvent));
-      return;
-    }
+    flicking.addPlugins(added.map(index => list[index]));
+    flicking.removePlugins(removed.map(index => prevList[index]));
+  }
+
+  private _getChildKeys() {
+    // Fill keys if not exist
+    const children = ([].slice.apply(this._cameraElRef.nativeElement.children) as HTMLElement[])
+      .filter(node => node.nodeType === Node.ELEMENT_NODE);
+    children.forEach(child => {
+      if (!(child as any).__NGX_FLICKING_KEY__) {
+        (child as any).__NGX_FLICKING_KEY__ =  uuid.v4();
+      }
+    });
 
-    // If Promise is not supported (IE)
-    if (this.criticalSection) {
-      setTimeout(() => {
-        // This works OK but it may cause blink when panel is appended or added
-        this.renderPanelChange.emit(renderChangeEvent);
+    return children.map(child => (child as any).__NGX_FLICKING_KEY__);
+  }
+
+  private _reRender = () => {
+    const flicking = this._nativeFlicking;
+    const visiblePanels = flicking.visiblePanels;
+
+    Promise.resolve().then(() => {
+      this.renderPanelChange.emit({
+        visibles: visiblePanels
+          .sort((panel1, panel2) => (panel1.position + panel1.offset) - (panel2.position + panel2.offset))
+          .map(panel => this.data[panel.index])
       });
-    } else {
-      this.renderPanelChange.emit(renderChangeEvent);
-    }
+    })
   }
 }
+
+export interface NgxFlickingComponent extends NativeFlicking {};
```

Comment:
There is an issue where angular or some compilers do not recognize the class.

So I created FlickingInterface.ts and inherited it.

Reply:
Is that an issue for the library users? I didn't have any issues while I'm building that.

Reply:
Yes. This will be a problem in angular's ivy compiler or product build.